### PR TITLE
gears.string doc revamp

### PR DESCRIPTION
**Pull request by SethBarberee**
PR fixes the following things:
* Document a missing parameter in `gears.string.quote_pattern`
* Add examples as requested for more of the gears module

Issues with PR at the moment:
* Can't get output to generate and display.. I don't know why mine don't display versus `gears.object`. 
(EDIT): As long as file exists, LDoc displays the output.

EDIT (12/17/19): Guess an Early Christmas gift, but I've added moar tests and tried to spice up the doc more. Would love reviews before I change this to a real PR.

### CODE REVIEW

**Comment by Aire-One** [comment]
[The travis build error](https://travis-ci.com/awesomeWM/awesome/jobs/267058151) is the same you should have in your local build (at least, I got it too while compiling your branch):

```
/home/travis/build/awesomeWM/awesome/build/lib/gears/string.lua:49: ?: id tag: 'name" has trailing text ; use not_luadoc=true if you want description to continue between tags

"

 

**Usage example output**:

"

/home/travis/build/awesomeWM/awesome/build/lib/gears/string.lua:70: ?: id tag: 'name" has trailing text ; use not_luadoc=true if you want description to continue between tags

"

 

**Usage example output**:

"

/home/travis/build/awesomeWM/awesome/build/lib/gears/string.lua:101: ?: id tag: 'name" has trailing text ; use not_luadoc=true if you want description to continue between tags

"

 

**Usage example output**:

"
``` 

This basically means you have some random ldoc tags within function description. Or rather, you are continuing the description after other ldoc tags.

**What's on your code?**

```lua
--- Count number of lines in a string.
-- @tparam string text Input string.
-- @treturn int Number of lines.
-- @staticfct gears.string.linecount
-- @DOC_text_gears_string_linecount_EXAMPLE@
```

The `@DOC_***_EXAMPLE@` tags are not managed by ldoc. They are part of the awesome wm "compiling" process. We have a script running before ldoc to paste the generated code of examples instead of their `@DOC***` tags. The generated code can be an image, code example, code output, ... depending on the first line parameter `--DOC_GEN_OUTPUT`, `--DOC_GEN_IMAGE`, ...

**So what's wrong with your code?**

The generated code is inserted as function description to be interpreted by ldoc. 

Ldoc comment works like this:

```lua
--- Summary of the function, not too long and with a dot at the end.
-- Description of the function. Can be as long as you want, and can
-- use _markdown_. (Actually you can use markdown everywhere...)
--
-- Use a blank line to generate a carriage return character in the description.
-- @ldoc-tags bla...
-- @ldoc-tags bla...
-- @ldoc-tags bla...
-- @ldoc-tags bla...
function foo()
    ...
end
```

With your current code, the description block continue after `@ldoc-tags`.

**How to fix it?**

Place the `@DOC*` lines at the function description body.

```lua
--- Count number of lines in a string.
-- @DOC_text_gears_string_linecount_EXAMPLE@
-- @tparam string text Input string.
-- @treturn int Number of lines.
-- @staticfct gears.string.linecount
```

(**NOTE:** You still can add a description before or after the `@DOC` tag. (description before examples is better IMHO but both style are possible))

I think your local documentation was generated before you wrote this PR. So all the pages were already there and you only see them not being modified (ldoc failed so apply no change).

I recommend you to run the `make distclean` target from time to time before compiling so you can see if you broke something easily.

Also, awesome makefile defines build targets to generate documentation. You can use `make ldoc` to run the documentation process only.

**Comment by Aire-One** [comment]
While images are generated, it seems text output aren't. The `example.output.txt` files used as results for text example are part of the git tree. In the commit b5ca8bf93779bd5fef55980f7608c0c9131ca362, some of them are manually added. After removing them, their examples broke and I can't find a build target to generate them again.

So I guess you have to write them manually to complete your examples. Which make me feels bad...

I'm maybe missing something, tho... CC @psychon

**Comment by SethBarberee** [comment]
@Aire-One adding manually the output worked. Mine aren't too complex so I added it. Also, did your fix for LDoc. Thanks for the detailed explanation!

File: tests/examples/text/gears/string/linecount.output.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Count is 1
+Count is 3
```
Comment (actionless) [inline tests/examples/text/gears/string/linecount.output.txt]:
```
Expected text from /home/travis/build/awesomeWM/awesome/tests/examples/text/gears/string/linecount.output.txt, but got:
--- /home/travis/build/awesomeWM/awesome/tests/examples/text/gears/string/linecount.output.txt	2019-12-16 20:24:43.819949295 +0000
+++ /tmp/tmp.aOArmfOLFH	2019-12-16 20:26:14.279949295 +0000
@@ -1,2 +1,2 @@
-Count is 1
-Count is 3
+Count is: 1
+Count is: 3
```

Comment (SethBarberee) [reply tests/examples/text/gears/string/linecount.output.txt]: Oops, nice catch

**Comment by SethBarberee** [comment]
Added 2 more tests for `startswith` and `endwith`. Also, tried to clean up the input/output parameter documentation a bit for the module

**Comment by Elv13** [comment]
Good progress!

The tests are currently failing because the output of what is in the .txt file no longer matches the test output.

File: tests/examples/text/gears/string/linewrap.lua [hunk 2]
```diff
@@ -0,0 +1,6 @@
+--DOC_GEN_OUTPUT --DOC_HIDE
+local gears = require("gears") --DOC_HIDE
+
+local test = "doit"
```
Comment (actionless) [inline tests/examples/text/gears/string/linewrap.lua]:
```suggestion
local test = "do it"
```

Comment (SethBarberee) [reply tests/examples/text/gears/string/linewrap.lua]: Realized it as soon as I went on family errands!

File: lib/gears/string.lua [hunk 3]
```diff
@@ -60,6 +62,9 @@ end
 --- Escape all special pattern-matching characters so that lua interprets them
 -- literally instead of as a character class.
 -- Source: http://stackoverflow.com/a/20778724/15690
+-- @DOC_text_gears_string_quote_pattern_EXAMPLE@
+-- @tparam s string to generate pattern for
```
Comment (Aire-One) [inline lib/gears/string.lua]:
```suggestion
-- @tparam string s String to generate pattern for.
```

File: lib/gears/string.lua [hunk 4]
```diff
@@ -68,7 +73,7 @@ function gstring.quote_pattern(s)
 end
 
 --- Generate a pattern matching expression that ignores case.
--- @param s Original pattern matching expression.
+-- @tparam s string Original pattern matching expression.
```
Comment (Aire-One) [inline lib/gears/string.lua]:
```suggestion
-- @tparam string q Original pattern matching expression.
```

File: lib/gears/string.lua [hunk 5]
```diff
@@ -102,16 +107,20 @@ function gstring.split(str, delimiter)
 end
 
 --- Check if a string starts with another string.
+-- @DOC_text_gears_string_startswith_EXAMPLE@
 -- @tparam string str String to search
+-- @treturn boolean true if string starts with specified string
```
Comment (Aire-One) [inline lib/gears/string.lua]:
```suggestion
-- @treturn boolean `true` if string starts with specified string.
```

File: lib/gears/string.lua [hunk 6]
```diff
@@ -102,16 +107,20 @@ function gstring.split(str, delimiter)
...
 -- @staticfct gears.string.startswith
 function gstring.startswith(str, sub)
     return string.sub(str, 1, string.len(sub)) == sub
 end
 
 --- Check if a string ends with another string.
+-- @DOC_text_gears_string_endswith_EXAMPLE@
 -- @tparam string str String to search
 -- @tparam string sub String to check for.
+-- @treturn boolean true if string ends with specified string
```
Comment (Aire-One) [inline lib/gears/string.lua]:
```suggestion
-- @treturn boolean `true` if string ends with specified string.
```

**Comment by SethBarberee** [comment]
Pretty happy with how it looks so I'll open it for review

**Comment by SethBarberee** [comment]
Squashed all commits into one commit

**Pull request merged**